pkg/controller/elasticsearch/services/services.go

```diff
@@ -90,3 +92,23 @@ func GetExternalService(c k8s.Client, es v1alpha1.Elasticsearch) (corev1.Service
 
 	return svc, nil
 }
+
+// ElasticsearchURL calculates the base url for Elasticsearch, taking into account the currently running pods.
+// If there is a HTTP scheme mismatch between spec and pods we switch to requesting individual pods directly
+// otherwise this delegates to ExternalServiceURL.
+func ElasticsearchURL(es v1alpha1.Elasticsearch, pods []corev1.Pod) string {
+	var schemeChange bool
+	for _, p := range pods {
+		scheme, exists := p.Labels[label.HTTPSchemeLabelName]
+		if exists && scheme != es.Spec.HTTP.Scheme() {
+			// scheme in existing pods does not match scheme in spec, user toggled HTTP(S)
+			schemeChange = true
+		}
+	}
+	if schemeChange {
+		// switch to sending requests directly to a random pod instead of going through the service
+		randomPod := pods[rand.Intn(len(pods))]
+		return fmt.Sprintf("%s://%s.%s:%d", randomPod.Labels[label.HTTPSchemeLabelName], randomPod.Name, randomPod.Namespace, network.HTTPPort)
```

Comment:
Got this error:
```
    --- FAIL: TestMutationHTTPToHTTPS/Add_some_data_to_the_cluster_before_starting_the_mutation (0.04s)
        require.go:794:
            	Error Trace:	steps_mutation.go:47
            	Error:      	Received unexpected error:
            	            	Put http://test-mutation-http-to-https-rzdh-es-masterdata-0.e2e-68x09-mercury:9200/data-integrity-check: dial tcp: lookup test-mutation-http-to-https-rzdh-es-masterdata-0.e2e-68x09-mercury on 10.96.0.10:53: no such host
            	Test:       	TestMutationHTTPToHTTPS/Add_some_data_to_the_cluster_before_starting_the_mutation
FAIL
FAIL	github.com/elastic/cloud-on-k8s/test/e2e/es	65.316s
```

IIUC the [statefulset documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id) we need a headless service between the Pod's name and the namespace to be able to resolve a POD name:

```
[elastic@e2e-68x09-mercury-ns-operator-0 ~]$ ping test-mutation-http-to-https-rzdh-es-masterdata-0.test-mutation-http-to-https-rzdh-es-masterdata.e2e-68x09-mercury
PING test-mutation-http-to-https-rzdh-es-masterdata-0.test-mutation-http-to-https-rzdh-es-masterdata.e2e-68x09-mercury.svc.cluster.local (10.244.2.4) 56(84) bytes of data.
64 bytes from 10-244-2-4.test-mutation-http-to-https-rzdh-es-http.e2e-68x09-mercury.svc.cluster.local (10.244.2.4): icmp_seq=1 ttl=62 time=0.077 ms
```

Reply:
I think you are right 🥇  Interestingly this error does not occur when running the e2e tests locally, with our port-forwarding tooling which is what I did 😞